src/network.js

```diff
@@ -142,7 +142,9 @@ class Network {
     const connectAttempts = []
     for await (const provider of this.findProviders(cid, CONSTANTS.maxProvidersPerRequest, options)) {
       this._log('connecting to providers', provider.id.toB58String())
-      connectAttempts.push(this.connectTo(provider, options))
+
+      this._libp2p.peerStore.addressBook.set(provider.id, provider.multiaddrs)
+      connectAttempts.push(this.connectTo(provider.id, options))
```

Comment:
Libp2p should be responsible for this. In the DHT we add discovered multiaddrs to the addressBook when we do `findPeer`, but apparently not when we do `findProviders`. In the delegated side of things, we don't add them at all.

So, I think we should follow exactly the same approach as with the `peer-discovery` modules, which libp2p [adds to the AddressBook](https://github.com/libp2p/js-libp2p/blob/v0.30.1/src/index.js#L606). In this case, we should add the multiaddrs to the peerStore on [findProviders](https://github.com/libp2p/js-libp2p/blob/v0.30.1/src/content-routing.js#L64), [findPeer](https://github.com/libp2p/js-libp2p/blob/v0.30.1/src/peer-routing.js#L94) and [getClosestPeers](https://github.com/libp2p/js-libp2p/blob/v0.30.1/src/peer-routing.js#L123). We can later remove some redundant addressBooks adds on the DHT, but we need to be careful with this as DHT queries might discover important peers to store that might not be returned in these APIs.

What do you think?

A side note, we should use `add` here as we might already know other multiaddrs of the peer that we can use in the following connect. Only `identify` should use `set` at this moment, as this is the source of truth

Reply:
I've added the storing of multiaddrs to findProviders here: https://github.com/libp2p/js-libp2p/pull/865 also made it a lot faster by returning results as they come in instead of waiting for one of the lookups to finish first.

I'll expand it to perform the same operations for findPeer and getClosestPeers tomorrow..